pytorch/torchtune
# Early fusion multimodal models

## Context
This enables Early Fusion models based on @pbontrager 's excellent original RFC on multimodal fusion models #1283. Since the RFC, we have already landed Deep Fusion model components. This PR discusses and implements the EarlyFusionModel component, along with testing and some lint updates.

Early fusion is simply a decoder with 1 or more extra encoders that merges their outputs with the token embeddings tables. The challenge lies in how we merge the embeddings and pass it into the decoder.

## Changelog
- Added EarlyFusionModel and tests
- Updated DeepFusionModel docstring, none of the code was touched except updated an incorrect typing
- Split `_fusion.py` into `_fusion_layers.py`, `_early_fusion.py`, and `_deep_fusion.py`
- Small update in peft utils

## Design
There is one design consideration I am seeking feedback on, and that is the EarlyFusionModel's usage of `self.decoder.tok_embeddings`. It accesses the decoder's token embedding table outside of the decoder forward because we need to merge the image encoder and any other modality encoder's output embeddings with the text embeddings (in this case just concatenate in sequence dimension):

```
embeds = self.tok_embeddings(tokens)
bsz, seq_len, embed_dim = embeds.shape
for encoder, inp in (encoder_input or {}).items():
    encoder_embeds = self.encoders[encoder](**inp)
    encoder_mask = (tokens == self.encoder_tokens[encoder]).expand(bsz, seq_len, embed_dim)
    embeds[encoder_mask] = encoder_embeds
    
output = self.decoder(embeds, mask, input_pos)
return output
```

Now, instead of token ids, we are passing in the merged embeddings directly into the decoder. But since we already used the text-only tok_embeddings from the decoder, we need to skip it when passing in the merged embeddings for the final decoder output. There are two ways we can do this.

#### State dict surgery
In the current code changes and suggested by the original RFC, we can manually set `self.decoder.tok_embeddings = nn.Identity()` so that it becomes a no-op when you forward pass with merged embeddings.
- This will require additional state dict hooks to make sure checkpoint saving and loading is still maintained despite the module change
- If a user wants to use the decoder outside of the EarlyFusionModule in the same script, they will need to restore the original tok_embeddings module from nn.Identity

#### Additional input_embeds kwarg
We could add a new keyword argument in `TransformerDecoder` forward for input embeddings. If this is passed in, we automatically skip the token embeddings:
```
h = self.tok_embeddings(tokens) if input_embeds is None else input_embeds
```
This way we don't need any state dict hooks or decoder modifications. However, we are polluting the decoder model forward with more arguments.

## Test plan
- Add forward pass decoder + encoders test
- Add forward pass decoder only test
- Add forward pass encoders only test, check to see embeddings are merged correctly in correct sequence placement
- Verify state dict hooks by loading a state dict with "decoder" labels then saving the same state dict
- Verify that variable number of encoder special tokens in each sample still works as expected
